src/SonarQube.Client/Services/RequestFactory.cs

```diff
@@ -0,0 +1,98 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SonarQube.Client
+{
+    public class RequestFactory
+    {
+        private readonly Dictionary<Type, SortedList<Version, Func<object>>> requestMappings =
+            new Dictionary<Type, SortedList<Version, Func<object>>>();
+
+        /// <summary>
+        /// Registers a simple request factory for the specified version of SonarQube. It works only for types that
+        /// have parameterless constructor.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request interface to create implementation for.</typeparam>
+        /// <typeparam name="TRequestImpl">The type implementing TRequest that has a parameterless constructor.</typeparam>
+        /// <param name="version">The version of SonarQube which first implements the request.</param>
+        /// <returns>Returns this.</returns>
+        public RequestFactory RegisterRequest<TRequest, TRequestImpl>(string version)
+            where TRequest : IRequest
+            where TRequestImpl : TRequest, new()
+        {
+            return RegisterRequest<TRequest, TRequestImpl>(version, () => new TRequestImpl());
+        }
+
+        /// <summary>
+        /// Registers a request factory for the specified version of SonarQube.
+        /// </summary>
+        /// <typeparam name="TRequest">The type of the request interface to create implementation for.</typeparam>
+        /// <typeparam name="TRequestImpl">The type implementing TRequest.</typeparam>
+        /// <param name="version">The version of SonarQube which first implements the request.</param>
+        /// <param name="factory">Factory function to create new instances of TRequestImpl.</param>
+        /// <returns>Returns this.</returns>
+        public RequestFactory RegisterRequest<TRequest, TRequestImpl>(string version, Func<TRequestImpl> factory)
+            where TRequest : IRequest
+            where TRequestImpl : TRequest
+        {
+            SortedList<Version, Func<object>> map;
+            if (!requestMappings.TryGetValue(typeof(TRequest), out map))
+            {
+                map = new SortedList<Version, Func<object>>();
+                requestMappings[typeof(TRequest)] = map;
+            }
+            map[Version.Parse(version)] = () => factory();
```

Comment:
We don't expect an entry to already exist, do we? If not, then it would be worth asserting that this is the case to help catch coding errors.